Mathlib/Tactic/TermCongr.lean

```diff
@@ -0,0 +1,695 @@
+to elaborate as an iff, eq, or heq.
+Later, the congruence generator handles any discrepencies.
+See `Mathlib.Tactic.TermCongr.CongrResult`. -/
+@[reducible, nolint unusedArguments]
+def cHole {α : Sort u} (val : α) {p : Prop} (_pf : p) : α := val
+
+/-- For error reporting purposes, make the hole pretty print as its value.
+We can still see that it is a hole in the info view on mouseover. -/
+@[app_unexpander cHole] def unexpandCHole : Lean.PrettyPrinter.Unexpander
+  | `($_ $val $_) => pure val
+  | _ => throw ()
+
+/-- Given a proposition, returns `(tyLhs, lhs, tyRhs, rhs)` if it's an eq, iff, or heq. -/
+def sides? (ty : Expr) : MetaM (Option (Expr × Expr × Expr × Expr)) := do
+  let ty ← whnf ty
+  if let some (lhs, rhs) := ty.iff? then
+    return some (.sort .zero, lhs, .sort .zero, rhs)
+  if let some (ty, lhs, rhs) := ty.eq? then
+    return (ty, lhs, ty, rhs)
+  else if let some (tyLhs, lhs, tyRhs, rhs) := ty.heq? then
+    return (tyLhs, lhs, tyRhs, rhs)
+  else
+    return none
+
+/-- Create the congruence hole. Used by `elabCHole`.
+
+Saves the current mvarCounter as a proxy for age. We use this to avoid
+reprocessing old congruence holes that happened to leak into the local context. -/
+def mkCHole (forLhs : Bool) (val pf : Expr) : MetaM Expr := do
+  -- Create a metavariable to bump the mvarCounter.
+  discard <| mkFreshTypeMVar
+  let d : MData := KVMap.empty
+    |>.insert congrHoleForLhsKey forLhs
+    |>.insert congrHoleIndex (← getMCtx).mvarCounter
+  return Expr.mdata d <| ← mkAppM ``cHole #[val, pf]
+
+/-- If the expression is a congruence hole, returns `(forLhs, sideVal, pf)`.
+If `mvarCounterSaved?` is not none, then only returns the hole if it is at least as recent. -/
+def cHole? (e : Expr) (mvarCounterSaved? : Option Nat := none) : Option (Bool × Expr × Expr) := do
+  match e with
+  | .mdata d e' =>
+    let forLhs : Bool ← d.get? congrHoleForLhsKey
+    let mvarCounter : Nat ← d.get? congrHoleIndex
+    if let some mvarCounterSaved := mvarCounterSaved? then
+      guard <| mvarCounterSaved ≤ mvarCounter
+    let #[_, val, _, pf] := e'.getAppArgs | failure
+    return (forLhs, val, pf)
+  | _ => none
+
+/-- Returns any subexpression that is a recent congruence hole.  -/
+def hasCHole (mvarCounterSaved : Nat) (e : Expr) : Option Expr :=
+  e.find? fun e' => (cHole? e' mvarCounterSaved).isSome
+
+/-- Eliminate all congruence holes from an expression by replacing them with their values. -/
+def removeCHoles (e : Expr) : Expr :=
+  e.replace fun e' => if let some (_, val, _) := cHole? e' then val else none
+
+/-- Elaborates a congruence hole and returns either the left-hand side or the right-hand side,
+annotated with information necessary to generate a congruence lemma. -/
+def elabCHole (h : Syntax) (forLhs : Bool) (expectedType? : Option Expr) : Term.TermElabM Expr := do
+  let pf ← Term.elabTerm h none
+  let pfTy ← inferType pf
+  unless ← isDefEq (← inferType pfTy) (.sort .zero) do
```

Comment:
Is this different from `(← inferType pfTy).isProp`?

Reply:
Yeah, it has the side effect of making `pfTy` a proposition if its type is just a metavariable. I'm adding "Ensure that `pfTy` is a proposition"